Libraries/LibCompress/Gzip.cpp

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibCompress/Gzip.h>
+
+#include <AK/String.h>
+
+namespace Compress {
+
+bool GzipDecompressor::BlockHeader::valid_magic_number() const
+{
+    return identification_1 == 0x1f && identification_2 == 0x8b;
+}
+
+bool GzipDecompressor::BlockHeader::supported_by_implementation() const
+{
+    if (compression_method != 0x08) {
+        // RFC 1952 does not define any compression methods other than deflate.
+        return false;
+    }
+
+    if (flags > Flags::MAX) {
+        // RFC 1952 does not define any more flags.
+        return false;
+    }
+
+    if (flags & Flags::FHCRC) {
+        TODO();
+    }
+
+    return true;
+}
+
+GzipDecompressor::GzipDecompressor(InputStream& stream)
+    : m_input_stream(stream)
+{
+}
+
+GzipDecompressor::~GzipDecompressor()
+{
+    if (m_state == State::ReadingDeflateBlock) {
+        m_block_stream.~DeflateDecompressor();
+        m_block_header.~BlockHeader();
+    }
+}
+
+// FIXME: Again, there are surely a ton of bugs because the code doesn't check for read errors.
+size_t GzipDecompressor::read(Bytes bytes)
+{
+    if (m_state == State::ReadingDeflateBlock) {
+        size_t nread = m_block_stream.read(bytes);
+
+        if (nread < bytes.size()) {
+            m_block_stream.~DeflateDecompressor();
+            m_state = State::Idle;
+
+            LittleEndian<u32> crc32, input_size;
+            m_input_stream >> crc32 >> input_size;
+
+            // FIXME: Validate crc32 and input_size.
```

Comment:
There is a CRC impl in LibCrypto/Checksum that you can use :)

Reply:
I've added it and it seems to work.